### PR TITLE
feat(chat): support function call api

### PR DESCRIPTION
**Pull request by Ccheers**
implement #360

### CODE REVIEW

**Comment by Ccheers** [comment]
@sashabaranov

File: chat.go [hunk 1]
```diff
@@ -43,12 +51,68 @@ type ChatCompletionRequest struct {
...
 	LogitBias        map[string]int          `json:"logit_bias,omitempty"`
 	User             string                  `json:"user,omitempty"`
+	Functions        []*FunctionsDefine      `json:"functions,omitempty"`
+	FunctionCall     string                  `json:"function_call,omitempty"`
 }
 
+type FunctionsDefine struct {
+	Name        string      `json:"name,omitempty"`
+	Description string      `json:"description,omitempty"`
+	Parameters  *FuncParams `json:"parameters,omitempty"`
```
Comment (vvatanabe) [inline chat.go:61]: @Ccheers The official documentation says `parameters` is optional, but when you actually call the API, it says `parameters` is required.

```
$ curl https://api.openai.com/v1/chat/completions -u :$OPENAI_API_KEY -H 'Content-Type: application/json' -d '{
  "model": "gpt-3.5-turbo-0613",
  "messages": [
    {"role": "user", "content": "What is the weather like in Boston?"}
  ],
  "functions": [
    {
      "name": "get_current_weather"
    }
  ]
}'
```

Response:
```json
{
   "error":{
      "message":"'parameters' is a required property - 'functions.0'",
      "type":"invalid_request_error",
      "param":null,
      "code":null
   }
}
```

Comment (Ccheers) [reply chat.go:61]: You are right and the type must be "object", but as a api we shouldn't limit it. Function `CreateChatCompletionWithFunctionCall` will do it

Comment (Ccheers) [reply chat.go:61]: we can add some comment to user

**Comment by zbysir** [comment]
Looking forward to it

**Comment by vvatanabe** [comment]
@sashabaranov @Ccheers  `CreateChatCompletionWithFunctionCall` function can aggregate multiple API calls related to "Function calling". Very useful. Excellent! 👍 👍 

But go-openai is just a client library for OpenAI API.  Should go-openai provide high-level abstraction like `CreateCompletionWithFunctionCall` function?

It may be difficult to design for abstraction between Azure OpenAI and OpenAI in the future when Azure OpenAI supports "Function calling".

I'm a bit concerned about other additional dependent libraries. Perhaps go-openai values maintenance cost and lightweight and dares to have zero dependencies.

It might be better to first just add some fields for "Function calling" to the `CreateChatCompletion` function request and response, and then gather user feedback to make a decision.

**Comment by zbysir** [comment]
> @sashabaranov @Ccheers  `CreateChatCompletionWithFunctionCall` function can aggregate multiple API calls related to "Function calling". Very useful. Excellent! 👍 👍 
> 
> But go-openai is just a client library for OpenAI API.  Should go-openai provide high-level abstraction like `CreateCompletionWithFunctionCall` function?
> 
> It may be difficult to design for abstraction between Azure OpenAI and OpenAI in the future when Azure OpenAI supports "Function calling".
> 
> I'm a bit concerned about other additional dependent libraries. Perhaps go-openai values maintenance cost and lightweight and dares to have zero dependencies.
> 
> It might be better to first just add some fields for "Function calling" to the `CreateChatCompletion` function request and response, and then gather user feedback to make a decision.

I quite agree! The signature of withfunctioncall is not universal enough, so users should decide for themselves.

**Comment by ushakov** [comment]
My 2p: it seems gjson library is only used to extract `choices.0.message."+fmt.Sprintf("args_%d", i)` field from JSON, which should be relatively easy to work around without introducing the additional dependency

**Comment by Ccheers** [comment]
> My 2p: it seems gjson library is only used to extract `choices.0.message."+fmt.Sprintf("args_%d", i)` field from JSON, which should be relatively easy to work around without introducing the additional dependency

It seems that using map[string]interface{} to deconstruct JSON for parameter extraction is also possible at present, but it is relatively more complicated.

**Comment by jmacwhyte** [comment]
I agree with the issues raised here, I think this library should be kept very simple.

Although the "function call" information that the API returns is meant to be used to call a function in the client's local environment, that might not be how everyone intends to use it. I would prefer to simply have the ChatCompletion/ChatCompletionStream structs updated to include to support the function call details, and leave it up to the client to decide how they will consume that information.

That said, I do think @Ccheers has done some great work which I will use as reference in my own project!

**Comment by Ccheers** [comment]
> > @sashabaranov @Ccheers  `CreateChatCompletionWithFunctionCall` function can aggregate multiple API calls related to "Function calling". Very useful. Excellent! 👍 👍
> > But go-openai is just a client library for OpenAI API.  Should go-openai provide high-level abstraction like `CreateCompletionWithFunctionCall` function?
> > It may be difficult to design for abstraction between Azure OpenAI and OpenAI in the future when Azure OpenAI supports "Function calling".
> > I'm a bit concerned about other additional dependent libraries. Perhaps go-openai values maintenance cost and lightweight and dares to have zero dependencies.
> > It might be better to first just add some fields for "Function calling" to the `CreateChatCompletion` function request and response, and then gather user feedback to make a decision.
> 
> I quite agree! The signature of withfunctioncall is not universal enough, so users should decide for themselves.

In the current implementation, JSON deserialization relies on Go Struct. However, the definition of `function call` seems to contradict our traditional deserialization approach. Therefore, I am considering extracting a separate function to support `function call`.

**Comment by Ccheers** [comment]
> I agree with the issues raised here, I think this library should be kept very simple.
> 
> Although the "function call" information that the API returns is meant to be used to call a function in the client's local environment, that might not be how everyone intends to use it. I would prefer to simply have the ChatCompletionMessage (etc) structs updated to include to support the function call details, and leave it up to the client to decide how they will consume that information.
> 
> That said, I do think @Ccheers has done some great work which I will use as reference in my own project!

I think this is a trade-off. If we consider introducing `function_call` into the native API, we may need to parse JSON twice in the native API, as discussed above. And not everyone needs `function_call`.

**Comment by jmacwhyte** [comment]
My vote is to not accept this PR just yet. We already have CreateChatCompletion() as well as CreateChatCompletionStream(), adding CreateChatCompletionWithFunctionCall() is a bit over the top when it's all the same API! I think the new function call feature should be able to fit nicely into the existing ChatCompletion structure (like @vvatanabe said), and I am willing to work on an alternate solution (tomorrow, probably) if no one else gets to it first.

**Pull request closed without merging**